cpp/daal/src/algorithms/k_nearest_neighbors/oneapi/cl_kernels/bf_knn_cl_kernels.cl

```diff
@@ -38,6 +38,26 @@ DECLARE_SOURCE(
         distances[global_id_0 + global_id_1 * N] = dataSq[global_id_0];
     }
 
+    __kernel void scatter_row_col(__global const algorithmFPType * dataSq, __global const algorithmFPType * querySq,
+                                  __global algorithmFPType * distances, int N) {
+        const int global_id_0 = get_global_id(0);
+        const int global_id_1 = get_global_id(1);
+
+        distances[global_id_0 + global_id_1 * N] = dataSq[global_id_0] + querySq[global_id_1];
+    }
+
+    __kernel void distances_from_squares(__global algorithmFPType * data) {
+        const int global_id_0     = get_global_id(0);
+        const algorithmFPType val = data[global_id_0];
+        data[global_id_0]         = (val > 0) ? sqrt(val) : 0;
```

Comment:
It looks like a workaround of the numerical error of the difference of squares. Moreover, the conditions for each thread are checked, which can take time. Is there any way we can work around this?

Reply:
I guess there is no such workaround except direct computation of distances.

Reply:
ok